### PR TITLE
Allow shared nested addons to be properly discovered.

### LINKED ISSUE #3240: Addon Discovery: Shared Nested Addons Cannot be Found
I am trying out ember-cli with ember@1.9.1 which needs handlebars@2. Unfortunately the broccoli-ember-hbs-template-compiler@1.7.1-alpha, which uses also handlebars@2, is not pushed to npmjs, so I had to install manually from source. To avoid conflicts, I removed the requirement from my application's package.json for broccoli-ember-hbs-template-compiler, since I am not managing it with npm.

So, the package is installed in node_modules, but not listed in package.json. In this situation, ember-cli warns with:

```
 DEBUG=express*,ember-cli:watcher* ember server --live-reload-port 35730 --port 4200
version: 0.1.12
  ember-cli:watcher detecting watchman +0ms
  ember-cli:watcher detected watchman: 3.0.0 +28ms
  ember-cli:watcher watchman 3.0.0 does satisfy: ^3.0.0 +9ms
Missing template processor
Error: Missing template processor
    at module.exports.preprocessTemplates (/home/user1/myapp/node_modules/ember-cli/lib/preprocessors.js:83:11)
    at EmberApp._processedTemplatesTree (/home/user1/myapp/node_modules/ember-cli/lib/broccoli/ember-app.js:523:10)
    at EmberApp.appAndDependencies (/home/user1/myapp/node_modules/ember-cli/lib/broccoli/ember-app.js:698:24)
    at EmberApp.javascript (/home/user1/myapp/node_modules/ember-cli/lib/broccoli/ember-app.js:758:34)
    at EmberApp.toArray (/home/user1/myapp/node_modules/ember-cli/lib/broccoli/ember-app.js:1041:10)
    at EmberApp.toTree (/home/user1/myapp/node_modules/ember-cli/lib/broccoli/ember-app.js:1063:30)
    at Object.<anonymous> (/home/user1/myapp/Brocfile.js:149:34)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
```

To solve the issue I had to enter again the broccoli-ember-hbs-template-compiler dependency in package.json.

It is misleading, because the module is already present in node_modules, and I was **sure** it was installed (since I had done it manually), so it took me a while to understand what was going on.

### PR DESCRIPTION
**Pull request by rwjblue**
Prior to these changes we always assumed that an addon was at the current package's (could be either the project or another addon) `node_modules/<addon-name>` path. This is true in many cases, but not all. Specifically, with more recent versions of NPM a package that is shared by a nested child package and an ancestor package will be deduped and used from that parent directory.

Simple example:

* Package A
  * Depends on Package B
  * Depends on Package C
* Package C depends on Package B

The folder structure in this case for more recent versions of NPM (2.0.0+) would be:

```
package-a/node_modules/package-b/*
package-a/node_modules/package-c/*
```

But our prior logic expected:

```
package-a/node_modules/package-b/*
package-a/node_modules/package-c/*
package-a/node_modules/package-c/node_modules/package-b/*
```

---

This PR ensures that we use the `resolve` package (which follows the [documented node resolution strategy](http://nodejs.org/api/modules.html#modules_all_together)) to determine the path to a given addon.

**Please note** that this PR also removes the ability to have an addon without any node-land JS (previously it was possible to have only a `package.json` and `vendor/` folder for example).  I am not aware of this being relied upon, but it is documented as a breaking change in the changelog for 0.2.0.

Closes #3240.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,18 @@
...
+Support for addon's without an entry point script (either `index.js` by default or the script specified by ember-addon main in the addon's `package.json`
+has been removed. An addon must have at least the following:
+
+```javascript
+module.exports = {
+  name: "addons-name-here"
+};
+```
+
+This should *not* pose a problem for the vast majority of addon's.
```
Comment (lukemelia) [inline CHANGELOG.md]: nit: remove the apostrophe

Comment (rwjblue) [inline CHANGELOG.md]: Good catch, updated.

**Comment by lukemelia** [comment]
Looks good to me. Thanks @rwjblue!

**Comment by rwjblue** [comment]
There are some acceptance test failures that I am working through...

**Comment by rwjblue** [comment]
Just confirmed that this fixes the demo repo (https://github.com/alvincrespo/test-new-app) provided by @alvincrespo in #3240.

**Comment by slindberg** [comment]
:beers:

**Pull request merged**